design/vrf-announcement.md

```diff
@@ -0,0 +1,198 @@
+to a VRF
+* Eventually, provide a guideline and an example on how to setup the VRF and the
+host routing, knowing that the routing logic might be CNI dependant
+
+### Non-Goals
+
+* Creating the VRF
+* Having MetalLB configure the host networking routing to drive the traffic
+towards / from the VRF
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I want to declare a BGP session specifying the host VRF to establish the session from.
+
+#### Story 2
+
+As a cluster administrator, I want to have the LoadBalancer IP announced to BGP peers connected from within a VRF
+
+## Design Details
+
+### VRF Design proposal
+
+FRR is able to run a bgpd process inside a VRF, so what needs to be done is to
+enable the specific configuration.
+From the [FRR docs](https://docs.frrouting.org/en/latest/bgp.html#clicmd-router-bgp-ASN-vrf-VRFNAME):
+
+```none
+router bgp ASN vrf VRFNAME
+
+    VRFNAME is matched against VRFs configured in the kernel. When vrf VRFNAME is not specified, the BGP protocol process belongs to the default VRF.
+```
+
+The proposal here is to define a new `vrf` field in the BGPPeer CR. By setting it,
+the speaker will instruct FRR to create a new router section
+related to the router and will configure the neighbour inside of it.
+
+```yaml
+apiVersion: metallb.io/v1beta2
+kind: BGPPeer
+metadata:
+  name: peer
+  namespace: metallb-system
+spec:
+  myASN: 64512
+  peerASN: 64512
+  peerAddress: 172.30.0.3
+  vrf: red
+```
+
+Any announcement meant to be performed towards the vrf-ed BGPPeer will be added
+in the related router's section.
+
+```none
+router bgp 100 vrf red
+  no bgp ebgp-requires-policy
+  no bgp network import-check
+  no bgp default ipv4-unicast
+
+  bgp router-id 10.1.1.255
+
+  neighbor 10.2.2.254 remote-as 200
+  neighbor 10.2.2.254 ebgp-multihop
+  neighbor 10.2.2.254 port 179
+  neighbor 10.2.2.254 timers 1 1
+  neighbor 10.2.2.254 password password
+  neighbor 10.2.2.254 update-source 10.1.1.254
+
+  address-family ipv4 unicast
+    neighbor 10.2.2.254 activate
+    neighbor 10.2.2.254 route-map 10.2.2.254-red-in in
+    network 172.16.1.10/24
+    neighbor 10.2.2.254 route-map 10.2.2.254-red-out out
+  exit-address-family
+```
+
+**Note**: We must be careful in handling the FRR configuration rendering,
+because all the route maps are handled in terms of the peer ip. With
+VRFs, we can have different BGP peers with the same IP.
+
+**Note 2**: There might be checks that we want to relax, like having
+BGPPeers with the same IP might be allowed in case the peers are reached
+from different VRFs.
+
+### How to configure host networking
+
+The right combination of host routing through ip rules and iptables rules is
+strongly dependant on the CNI. Because of this reason, we don't think is safe
+to have a one size-fit all solution. What we will provide with this change is
+a sample script (and possibly a controller) that works for kindnet (in order
```

Comment:
nit: works for the Kind network

Reply:
https://www.tkng.io/cni/kindnet/ 
Kindnet is the CNI used by kind